### PR TITLE
728 read secrets from azure key vault

### PR DESCRIPTION
**Pull request by felixtheflex**
### Context

Integrate terraform with keyvault

This is a move away from storing application and infrastructure secrets on Github Actions.

### Changes proposed in this pull request

data azurerm_key_vault

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -78,27 +78,36 @@ jobs:
...
-        working-directory: terraform
-        env:
-          ARM_ACCESS_KEY: ${{ secrets[format('ARM_ACCESS_KEY_{0}', env.DEPLOY_ENV)] }}
-          TF_VAR_paas_app_docker_image: ${{env.DOCKER_REPO}}:${{ github.event.inputs.sha }}
-
-      - name: Terraform Plan and Apply
-        run: |
           terraform plan -var-file=workspace-variables/${{ env.DEPLOY_ENV }}.tfvars
-          terraform apply -var-file=workspace-variables/${{ env.DEPLOY_ENV }}.tfvars -auto-approve
+          #terraform apply -var-file=workspace-variables/${{ env.DEPLOY_ENV }}.tfvars -auto-approve
```
Comment (saliceti) [inline .github/workflows/deploy.yml:105]: Commented for debugging?

File: Makefile [hunk 2]
```diff
@@ -66,11 +101,14 @@ destroy: terraform-init
 	cd terraform && terraform destroy -var-file=workspace-variables/$(env_config).tfvars
 
 terraform-init:
-	cd terraform && terraform init -reconfigure -backend-config=workspace-variables/$(env_config)_backend.tfvars $(backend_key)
+	$(eval export TF_DATA_DIR=./terraform/.terraform)
```
Comment (saliceti) [inline Makefile:104]: What is this for?

File: bin/fetch_config.rb [hunk 3]
```diff
@@ -0,0 +1,346 @@
+#!/usr/bin/env ruby
```
Comment (saliceti) [inline bin/fetch_config.rb:1]: This file should not be committed in this repository

File: terraform/modules/statuscake/main.tf [hunk 4]
```diff
@@ -8,4 +8,5 @@ resource statuscake_test alert {
   trigger_rate   = each.value.trigger_rate
   confirmations  = 1
   node_locations = each.value.node_locations
+  paused         = each.value.paused
```
Comment (saliceti) [inline terraform/modules/statuscake/main.tf:11]: We agreed to not make this variable?

File: terraform/workspace-variables/staging.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export TF_VAR_key_vault_name="s121t01-shared-kv-01"
```
Comment (saliceti) [inline terraform/workspace-variables/staging.sh:1]: Quotes are not required

**Comment by felixtheflex** [comment]
[Read secrets from Azure key vault](https://trello.com/c/MuzYT0PH/728-read-secrets-from-azure-key-vault)

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -78,27 +78,36 @@ jobs:
           APP_SECRETS:   ${{ secrets[format('APP_SECRETS_{0}', env.DEPLOY_ENV)] }}
           INFRA_SECRETS: ${{ secrets[format('INFRA_SECRETS_{0}', env.DEPLOY_ENV)] }}
 
+      - uses: azure/login@v1
+        with:
+          creds: ${{ secrets[format('AZURE_CREDENTIALS_{0}', env.DEPLOY_ENV)] }}
+
+      - name: Validate Azure key vault secrets
```
Comment (vigneshmsft) [inline .github/workflows/deploy.yml:85]: can you use action from https://github.com/DFE-Digital/github-actions/tree/master/validate-key-vault-secrets
usage https://github.com/DFE-Digital/find-teacher-training/compare/master..103b3054fca7523d14511bc3c8558aa9647ce061#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3R66-R91

File: terraform/workspace-variables/staging.tfvars [hunk 7]
```diff
@@ -11,6 +11,10 @@ paas_worker_app_instances        = 1
 paas_worker_app_memory           = 512
 paas_worker_app_stopped          = false
 
+#KeyVault
+key_vault_resource_group    = "s121t01-shared-01"
```
Comment (vigneshmsft) [inline terraform/workspace-variables/staging.tfvars:15]: I believe this is incorrect, it should have a -`-rg` suffix ?

File: Makefile [hunk 8]
```diff
@@ -37,27 +45,54 @@ qa:
...
+		&& curl -s https://raw.githubusercontent.com/DFE-Digital/bat-platform-building-blocks/master/scripts/fetch_config/fetch_config.rb -o bin/fetch_config.rb \
+		&& chmod +x bin/fetch_config.rb \
+		|| true
+
+set-azure-account:
+	az account set -s ${AZ_SUBSCRIPTION}
 
+edit-app-secrets: install-fetch-config set-azure-account
+	. terraform/workspace-variables/$(env).sh && bin/fetch_config.rb -s azure-key-vault-secret:$${TF_VAR_key_vault_name}/$${TF_VAR_key_vault_app_secret_name} \
+		-e -d azure-key-vault-secret:$${TF_VAR_key_vault_name}/$${TF_VAR_key_vault_app_secret_name} -f yaml
```
Comment (saliceti) [inline Makefile:80]: Please add `-c` to ask for confirmation, as in https://github.com/DFE-Digital/bat-platform-building-blocks/blob/master/scripts/fetch_config/fetch_config.rb#L40
Same for edit-infra-secrets

File: .github/workflows/deploy.yml [hunk 9]
```diff
@@ -69,36 +74,36 @@ jobs:
...
 
       - name: Pin Terraform version
         uses: hashicorp/setup-terraform@v1.2.1
         with:
           terraform_version: 0.13.5
 
-      - name: Terraform init
+      - name: Terraform init, Terraform Plan and Apply
         run: |
+          . "workspace-variables/${DEPLOY_ENV}.sh"
```
Comment (vigneshmsft) [inline .github/workflows/deploy.yml:97]: this is not required here now..

File: terraform/variables.tf [hunk 10]
```diff
@@ -32,26 +32,32 @@ variable dockerhub_password { default = null }
 
 variable paas_app_config_file { default = "workspace-variables/app_config.yml" }
 
-variable paas_app_secrets_file { default = "workspace-variables/app_secrets.yml" }
-
-variable infra_secrets_file { default = "workspace-variables/infra_secrets.yml" }
-
```
Comment (vigneshmsft) [inline terraform/variables.tf:38]: also remove `paas_user, paas_password, dockerhub_username, dockerhub_password` variables

**Review by saliceti — APPROVED** [review]
LGTM

**Review by vigneshmsft — APPROVED** [review]

**Pull request merged**